config/config.go

```diff
@@ -186,19 +195,46 @@ func FromViper(v *viper.Viper) (*Config, error) {
 		cfg.Walk = walk.Stdin.String()
 	}
 
+	// set git-based tree root command if the walker is Git
+	if cfg.Walk == walk.Git.String() {
+		cfg.TreeRootCmd = "git rev-parse --show-toplevel"
+	}
+
 	// determine the tree root
+	//nolint:nestif
 	if cfg.TreeRoot == "" {
-		// if none was specified, we first try with tree-root-file
+		// at the treefmt command level we ensure that `--tree-root-file` and `--tree-root-cmd` are mutually exclusive
 		if cfg.TreeRootFile != "" {
 			// search the tree root using the --tree-root-file if specified
 			_, cfg.TreeRoot, err = FindUp(cfg.WorkingDirectory, cfg.TreeRootFile)
 			if err != nil {
 				return nil, fmt.Errorf("failed to find tree-root based on tree-root-file: %w", err)
 			}
-		} else {
-			// otherwise fallback to the directory containing the config file
-			cfg.TreeRoot = filepath.Dir(v.ConfigFileUsed())
 		}
+
+		if cfg.TreeRootCmd != "" {
+			// use the output of the tree root command as the tree root
+			parts := strings.Fields(cfg.TreeRootCmd)
+
+			ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+			defer cancel()
+
+			//nolint:gosec
+			cmd := exec.CommandContext(ctx, parts[0], parts[1:]...)
+			cmd.Dir = cfg.WorkingDirectory
+
+			out, cmdErr := cmd.CombinedOutput()
+			if cmdErr != nil {
+				return nil, fmt.Errorf("failed to run tree-root-cmd: %w", cmdErr)
+			}
+
+			cfg.TreeRoot = strings.TrimSpace(string(out))
+		}
+	}
+
+	if cfg.TreeRoot == "" {
```

Comment:
Do we want this to happen if `tree-root-cmd` successfully prints empty string? Wouldn't that be confusing for our users?

Reply:
Re-ordered the logic to account for this.

Reply:
I don't see how this has changed, sorry. If I use something like `--tree-root-cmd="echo -n ''"`, won't that command successfully print `""`, and then we'll go into this `if` statement?

Reply:
I hallucinated a change which I didn't actually commit. Refined it a bit more there to make the logic explicity more clearly mutually exclusive.